### PR TITLE
RFR: Improve equality operator for custom dicom objects

### PR DESCRIPTION
**Pull request by Centrus007**

### CODE REVIEW

**Comment by Centrus007** [comment]
This passed pytest and built the docs locally just fine...?

**Comment by Centrus007** [comment]
Ah, I imported packaging. Weird, I was under the impression that this normally comes with setuptools by default..

File: packages/pymedphys_dicom/tests/test_dicom_base.py [hunk 1]
```diff
@@ -83,4 +86,35 @@ def test_to_and_from_file():
...
+    assert dicom1_from_file == dicom2_from_file
+
+    dicom1_from_file.dataset.PatientName = 'test^PatientName change'
+    assert dicom1_from_file != dicom2_from_file # Negative case
+
+    dicom1_from_file.dataset.PatientName = 'Python^Monte'
+    assert dicom1_from_file == dicom2_from_file # Equality post re-assignment
+
+    dicom1_from_file_copied = deepcopy(dicom1_from_file)
+    assert dicom1_from_file == dicom1_from_file_copied # Equality from deepcopy
```
Comment (SimonBiggs) [inline packages/pymedphys_dicom/tests/test_dicom_base.py]: Replace the above with:

```python
    try:
        fp1 = DicomBytesIO()
        dicom1.to_file(fp1)
        fp2 = DicomBytesIO()
        dicom2.to_file(fp2)

        dicom1_from_file = DicomBase.from_file(fp1)
        dicom2_from_file = DicomBase.from_file(fp2)
        
        # Equality from file (implicitly also from dataset)
        assert dicom1_from_file == dicom2_from_file

        dicom1_from_file.dataset.PatientName = 'test^PatientName change'
        assert dicom1_from_file != dicom2_from_file  # Negative case

        dicom1_from_file.dataset.PatientName = 'Python^Monte'
        assert dicom1_from_file == dicom2_from_file  # Equality post re-assignment

        dicom1_from_file_copied = deepcopy(dicom1_from_file)
        assert dicom1_from_file == dicom1_from_file_copied  # Equality from deepcopy
    finally:
        fp1.close()
        fp2.close()
```

I think potentially those DicomBytesIO's stick around unless they're closed?

File: packages/pymedphys_dicom/src/pymedphys_dicom/dicom/_level3/dicom_collection.py [hunk 2]
```diff
@@ -61,25 +62,56 @@ def __init__(self, dataset, copy=True):
...
+        return self.dataset == other.dataset
+
+
     def __repr__(self):
         return self.dataset.__repr__()
 
+
     def __eq__(self, other):
-        return self.dataset.__eq__(other.dataset)
+        if version.parse(pydicom.__version__) <= version.parse("1.2.1"):
```
Comment (SimonBiggs) [inline packages/pymedphys_dicom/src/pymedphys_dicom/dicom/_level3/dicom_collection.py]: ahh, neat, better than pinning :). I like it!

**Comment by SimonBiggs** [comment]
Ahh I didn't see `tox` wasn't having fun. `tox` explicitly strips the environment to a bare minimum so that our dependency definitions are correct.

Just need to add `packaging` to our dependencies for `pymedphys-dicom`.

**Comment by Centrus007** [comment]
> Just need to add packaging to our dependencies for pymedphys-dicom.

Tried that in last push. Hopefully all good now :)

**Comment by SimonBiggs** [comment]
Also, running `tox -e py37` will run that tox command locally.

**Comment by Centrus007** [comment]
> Also, running tox -e py37 will run that tox command locally.

Cool! Good tip :)

**Pull request merged**